### DISCUSSION TITLE
how to make "optional" variant of a schema.

### DISCUSSION THREAD

**Discussion by drortirosh** [json-schema-org/community#283] [General]
I have an object with 5 fields.
But in some APIs (e.g. in a search API) I want declare the same ObjetType, but with some of the field optional.

How do I declare these two types?

**Comment by handrews** [comment]
Since [JSON Schema is a constraint system](https://modern-json-schema.com/json-schema-is-a-constraint-system), you start by defining the schema that is all-optional.  Then you can `"$ref"` it, and alongside that reference, use `"required"` to make some or all properties required.

In JSON Schema, you have to start with the most permissive version of a schema as the basis of re-use.  The more restrictive versions build on that.  This is because you cannot lift a constraint once it is applied.
